### PR TITLE
Allow applying kapt and using mixins in other source sets

### PR DESCRIPTION
**Pull request by natanfudge**
This PR fixes the following issues regarding refmaps: 
- Applying kapt makes you unable to build the project, see https://github.com/FabricMC/fabric-loom/issues/209.
- Mixin would not produce refmaps for source sets other than `main`. 
- There was a weird warning regarding refmaps. It is now only logged as `info`. 

Note: This PR might allow Kotlin mixins but that option is not supported.

### CODE REVIEW

**Review by dblsaiko — APPROVED** [review]
LGTM

Although I would make AnnotationProcessorInvoker<T> and everything extending it be static methods instead of classes since it looks like they get allocated, then you call configureMixin on them, and then they get immediately deleted and not stored anywhere, as far as I can see. But really that's just personal preference.

File: src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+				// Pass named MC classpath to mixin AP classpath
+				processorConfig.extendsFrom(
+								configs.getByName(Constants.MINECRAFT_NAMED),
+								configs.getByName(Constants.MOD_COMPILE_CLASSPATH_MAPPED),
+								configs.getByName(Constants.MAPPINGS_FINAL)
+				);
+
+				// Add Mixin and mixin extensions (fabric-mixin-compile-extensions pulls mixin itself too)
+				project.getDependencies().add(processorConfig.getName(),
+								"net.fabricmc:fabric-mixin-compile-extensions:" + Constants.MIXIN_COMPILE_EXTENSIONS_VERSION);
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:104]: I know im still against kotlin mixins, but is this really needed on the kotlin annotation processor? Does kapt work without this.

File: src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+		for (T task : invokerTasks) {
+			passMixinArguments(task);
+		}
+	}
+
+	static Stream<SourceSet> getNonTestSourceSets(Project project) {
+		return project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets()
+						.stream()
+						.filter(sourceSet -> !sourceSet.getName().equals("test"));
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:116]: Why is this needed? Should it be excluding custom test source sets like `testMod` in fabric api?

**Review by modmuss50 — CHANGES_REQUESTED** [review]
This will need some testing, need to make sure the old issues in idea dont come back with this.

File: src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java (hunk 1 above)
Comment (natanfudge) [reply src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:104]: When kapt it applied it invokes java mixins instead of the java plugin, so it needs to have mixin and the mc stuff on its classpath

Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:104]: Ok, thats what I assumed.

File: src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java (hunk 2 above)
Comment (natanfudge) [reply src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:116]: There is similar code in the old code that excludes sourcesets that include the work `test` in them, so I thought I would do the same (I assume the thought was that there's no point to applying mixin on the test sourceset). I made it only target specifically `test` though because having mixins arbitraily not work because your sourceset has the word `test` in them (like `testMod`) doesn't make sense.

Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:116]: Ok, that sounds resonable. As they are needed on the testMod sourceset so that makes sense. Does it cause any harm if they are ran on the test source set?

Comment (natanfudge) [reply src/main/java/net/fabricmc/loom/util/mixin/AnnotationProcessorInvoker.java:116]: No, it doesn't cause any issues. This code is just here to make it more similar to the intention of the old behavior.

File: src/main/java/net/fabricmc/loom/util/mixin/KaptApInvoker.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import kotlin.Unit;
```
Comment (liach) [inline src/main/java/net/fabricmc/loom/util/mixin/KaptApInvoker.java:35]: Hmm, is kotlin available in gradle environment?

Comment (natanfudge) [reply src/main/java/net/fabricmc/loom/util/mixin/KaptApInvoker.java:35]: Note that imports don't exist in JVM bytecode; kotlin.Unit is only accessed after an if check made earlier `hasPlugin("org.jetbrains.kotlin.kapt")`, and kapt plugin pulls Kotlin with it.

Comment (liach) [reply src/main/java/net/fabricmc/loom/util/mixin/KaptApInvoker.java:35]: Just concerned of `ClassNotFoundException` at runtime, asking for a safety check. Should be good since the method (or even this class) theoretically won't be loaded if the kapt plugin was not deteced

**Review by liach — APPROVED** [review]

**Review by Juuxel — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]
Tested and seems to not break things.
